src/Covenant/Internal/Data.hs

```diff
@@ -0,0 +1,151 @@
+isRecursiveChildOf :: TyName -> ValT AbstractTy -> Reader ScopeBoundary Bool
+isRecursiveChildOf tn = \case
+  Datatype tn' args
+    | tn' == tn -> checkArgsIsRec 0 (V.toList args)
+    | otherwise -> pure False
+  _ -> pure False
+  where
+    -- Checks that the arguments to a datatype (which we expect will be checked to have the same tycon as the parent type)
+    -- have the correct DeBruijn index & the same order as in the initial binding context of the declaration.
+    -- Has to be monadic because need access to the current scope level.
+    checkArgsIsRec :: Int -> [ValT AbstractTy] -> Reader ScopeBoundary Bool
+    checkArgsIsRec _ [] = pure True
+    checkArgsIsRec n (x : xs) = case x of
+      Abstraction (BoundAt db varIx) -> do
+        ScopeBoundary here <- ask
+        let dbInt = asInt db
+        -- Explanation: A component ValT is only a recursive instance of the parent type if
+        --              the DeBruijn index of its type variables points to Z (and the other conditions obtain)
+        if dbInt - here == 0 && review intIndex varIx == n
+          then checkArgsIsRec (n + 1) xs
+          else pure False
+      _ -> pure False
+
+allComponentTypes :: DataDeclaration AbstractTy -> [ValT AbstractTy]
+allComponentTypes = toListOf (datatypeConstructors % folded % constructorArgs % folded)
+
+-- This tells us whether a ValT *contains* a direct recursive type. I.e it tells us whether we need to construct a base functor
+hasRecursive :: TyName -> ValT AbstractTy -> Reader ScopeBoundary Bool
+hasRecursive tn = \case
+  Abstraction {} -> pure False
+  BuiltinFlat {} -> pure False
+  -- NOTE: This assumes that we've forbidden higher rank arguments to constructors (i.e. we can ignore the scope here)
+  ThunkT (CompT _ (CompTBody (NEV.toList -> compTArgs))) -> local (+ 1) $ do
+    or <$> traverse (hasRecursive tn) compTArgs
+  dt@(Datatype _ args) -> do
+    thisTypeIsRecursive <- isRecursiveChildOf tn dt
+    aComponentIsRecursive <- or <$> traverse (hasRecursive tn) args
+    pure $ thisTypeIsRecursive || aComponentIsRecursive
+
+-- Yeah yeah I could use ReaderT but this is less awkward imo, we only care about the "maybe" at the top level when detecting recursion
+mkBaseFunctor :: DataDeclaration AbstractTy -> Reader ScopeBoundary (Maybe (DataDeclaration AbstractTy))
+mkBaseFunctor (DataDeclaration tn numVars ctors) = do
+  anyRecComponents <- or <$> traverse (hasRecursive tn) allCtorArgs
+  if null ctors || not anyRecComponents
+    then pure Nothing
+    else do
+      baseCtors <- traverse mkBaseCtor ctors
+      pure . Just $ DataDeclaration baseFName baseFNumVars baseCtors
+  where
+    -- TODO: I think we were going to make this "illegal" so users can't use it directly but I forget the legality rules
+    baseFName = case tn of
+      TyName tyNameInner -> TyName (tyNameInner <> "_F")
+
+    baseFNumVars :: Count "tyvar"
+    baseFNumVars = fromJust . preview intCount $ review intCount numVars + 1
```

Comment:
You can do `numVars <> count1` here I think.

Reply:
No semigroup instance for `Count` (though there is one for `Index`).